### PR TITLE
🐞fix support for non-square images

### LINKED ISSUE #136: GANomaly breaks when changing image size
**Describe the bug**
- GANomaly training fails when changing the image size in the config.

**To Reproduce**
1. Change the `image_size` parameter in the Patchcore config, to for example `[256, 234]`
2. Run training.
3. Training will fail with a runtime error.

**Expected behavior**
- Training should succeed, or notify the user that the specified image size is invalid.

**Hardware and Software Configuration**
- OS: Ubuntu
- NVIDIA Driver Version [470.103.01]
- CUDA Version [e.g. 11.1]

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Description

- Computes padding based on the image size. Might not be the best approach.

- Fixes #136 

## Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/ganomaly/torch_model.py [hunk 1]
```diff
@@ -40,13 +42,14 @@ def __init__(
     ):
         super().__init__()
 
-        assert input_size % 16 == 0, "Input size should be a multiple of 16"
+        assert input_size[0] % 16 == 0 and input_size[1] % 16 == 0, "Input size should be a multiple of 16"
```
Comment (djdameln) [inline anomalib/models/ganomaly/torch_model.py:45]: I guess this is no longer needed now that we use padding to make sure that the network can process any image size.

File: anomalib/models/ganomaly/torch_model.py [hunk 2]
```diff
@@ -74,14 +77,33 @@ def __init__(
...
+
+        Args:
+            input_size (Tuple[int, int]): Input size
+
+        Returns:
+            Tuple[int, int]: Padding for each dimension
+        """
+        # find the largest dimension
+        l_dim = 2 ** math.ceil(math.log(max(*input_size), 2))
+        padding = math.ceil((l_dim - input_size[0]) // 2 + 1), math.ceil((l_dim - input_size[1]) // 2 + 1)
```
Comment (djdameln) [inline anomalib/models/ganomaly/torch_model.py:104]: This won't work for odd image size, e.g. 221

**Review by djdameln — CHANGES_REQUESTED** [review]

File: anomalib/models/ganomaly/torch_model.py [hunk 3]
```diff
@@ -363,6 +394,26 @@ def forward(self, batch: Tensor) -> Tensor:
         Returns:
             Tensor: Regeneration scores.
         """
+        padded_batch = self.get_padded_tensor(batch)
         self.generator.eval()
-        _, latent_i, latent_o = self.generator(batch)
+        _, latent_i, latent_o = self.generator(padded_batch)
         return torch.mean(torch.pow((latent_i - latent_o), 2), dim=1).view(-1)  # convert nx1x1 to n
+
+    def get_padded_tensor(self, batch: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/ganomaly/torch_model.py:402]: I'll update it in a bit. Static is a good idea. In that case would it make sense to move it to `data/utils/image.py`?

[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/models/ganomaly/torch_model.py:402]: Yes, I think it would make sense to move it to a shared location so that other modules could access it as well. In that case we should maybe name it `pad_nextpow2` or similar

File: anomalib/data/utils/image.py [hunk 4]
```diff
@@ -66,3 +69,23 @@ def read_image(path: Union[str, Path]) -> np.ndarray:
     image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
 
     return image
+
+
+def pad_nextpow2(batch: Tensor) -> Tensor:
+    """Compute required padding from input size and return padded images.
+
+    Finds the largest dimension and computes a square image to pass into the model. In case the image dimension
```
Comment (djdameln) [inline anomalib/data/utils/image.py:77]: Maybe mention in the description that the function pads the images so that both sides are a power of 2.

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks! Just one minor comment left

**Review by djdameln — APPROVED** [review]
